### PR TITLE
feat: add user specific notification settings

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

The settings can be customized for each origin and each processor individually. The default Web indicates notifications shown in the Backstage UI. By default, if there are no settings saved in the database, all notifications are enabled for all processors.

The origins will populate by time for each user as they receive the first notification from that origin. Processors are shown as their own columns.

Later, if it makes sense, allow users to also disable/enable notifications based on notification topic.

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend/sample-templates/all-templates.yaml
Comment (benjdlambert) [inline plugins/scaffolder-backend/sample-templates/all-templates.yaml:6]: I'm wondering if it makes sense to push these examples to the `examples` option when creating the action and then people get them for free everywhere in the actions view rather than having to dig around for this template? :pray:

**Review by benjdlambert — COMMENTED** [review]
Quick comment about the `scaffolder-backend` changes

File: plugins/scaffolder-backend/sample-templates/all-templates.yaml
Comment (drodil) [reply plugins/scaffolder-backend/sample-templates/all-templates.yaml:6]: Hey, yes, those are in the examples available here https://github.com/drodil/backstage/blob/6fc03c7d3ab0b1ff75ea037c113314bf47fecb99/plugins/scaffolder-backend-module-notifications/src/actions/sendNotification.examples.ts#L20

Adding this template mainly allows easier debugging of notifications and allows them to be seen in action when running the local development environment. Especially at the moment, where none of the core features yet use notifications. Of course, the notifications can be done with `curl` or a similar direct API call but it's a bit of a hassle every time.

File: plugins/notifications-backend/src/service/router.ts
Comment (mareklibra) [inline plugins/notifications-backend/src/service/router.ts:92]: It took me a while to understand what a `channel` means. In fact it's a list of processors plus the standard UI (newly called as Web).

I would stick to the `processors` wording. If not then at least clearly describe the meaning for new comers.

**Review by mareklibra — COMMENTED** [review]
Works nice, I would just change the wording. In fact it's a configuration (disable/enable) for processors per user.

I like the idea of introducing the Settings' tab `Notifications` as we can add more there later, e.g. tweak the UI defaults.

In the context of this PR, I would clearly mark, that the settings section is for processors.

File: plugins/notifications-backend/src/service/router.ts
Comment (drodil) [reply plugins/notifications-backend/src/service/router.ts:92]: Yeah maybe makes sense, have to think about this again - I think we had discussion about this already with @Rugvip during the BEP review... Was there some conclusion? I can't remember 😄

Comment (drodil) [reply plugins/notifications-backend/src/service/router.ts:92]: Related part in the BEP: https://github.com/backstage/backstage/tree/master/beps/0001-notifications-system#user-specific-notification-settings

**Comment by mareklibra** [comment]
I have no code-wise comments, just the naming.

I would avoid using `channel` for processors (more or less). My motivation is in increasing readability for future contributors. @drodil , wdyt?

In case I am the only one, I will not block merging this. It lgtm otherwise.

**Comment by drodil** [comment]
@mareklibra I would prefer `channel` over `processor` as the Web is not a processor (though it could have been if we had gone that way). Later we probably need to extend this to some global settings and also include topics but that can be done later as this is on enough level for the work. @Rugvip any comments from your side?

**Comment by drodil** [comment]
This probably still waiting for @Rugvip to get back, but otherwise, I rebased and fixed the merge conflicts so this is ready for review+merge!

**Comment by drodil** [comment]
Any chance to get this included in the upcoming release?

**Comment by mareklibra** [comment]
I have no more comments except the naming. Let's not block the feature just on that.

**Review by mareklibra — APPROVED** [review]

**Comment by drodil** [comment]
Not stale, needs a rebase though. @Rugvip any chance to get this forward?

**Comment by drodil** [comment]
Not stale, waiting for final review & merge

**Comment by mareklibra** [comment]
There are no additional comments so far.
@drodil If you rebase it, I will merge it.

**Comment by drodil** [comment]
@mareklibra ok, rebased though the Verify Docs Quality job fails due to missing python. Not related failure though.

**Comment by mareklibra** [comment]
Re-run helped.

**Comment by mareklibra** [comment]
Seems I do not have all privileges to merge it.
Cc @benjdlambert @freben

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts
Comment (benjdlambert) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:465]: Think it might be worthwhile using `this.db<RowType>` here to make these types easier to use.

File: plugins/notifications-backend/src/database/NotificationsStore.ts
Comment (benjdlambert) [inline plugins/notifications-backend/src/database/NotificationsStore.ts:88]:
```suggestion
  getUserNotificationOrigins(options: { user: string }): Promise<{ origins: string[] }>;
```

Comment (benjdlambert) [reply plugins/notifications-backend/src/database/NotificationsStore.ts:88]: might not to too much of a big deal as I think that these are private right?

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Hey :wave: sorry for the delay here.

Few things just stood out at a glance here, gonna bring in @Rugvip for another set of eyes too.

**Review by Rugvip — APPROVED** [review]
Alright, awesome stuff and sorry for moving slowly with one!

Let's :shipit: for tomorrow's next release

**Pull request merged**
